src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -817,167 +817,256 @@ private BoundCollectionExpression ConvertCollectionExpression(
+                        : ErrorCode.WRN_CollectionExpressionRefStructMayAllocate,
+                        syntax, _targetType);
+                }
 
-                case CollectionExpressionTypeKind.CollectionBuilder:
-                    {
-                        var namedType = (NamedTypeSymbol)targetType;
+                // Specialized handling for ienumerable-based normal collections.  These defer to the behavior we had
+                // since C# 3.0 where we will determine which .Add methods to call on an element by element basis.
+                if (collectionTypeKind == CollectionExpressionTypeKind.ImplementsIEnumerable)
+                    return TryConvertCollectionExpressionImplementsIEnumerableType(constructor);
 
-                        var collectionBuilderMethods = GetCollectionBuilderMethods(syntax, namedType, diagnostics, forParams: false);
-                        if (collectionBuilderMethods.IsEmpty)
-                        {
-                            return BindCollectionExpressionForErrorRecovery(node, targetType, inConversion: true, diagnostics);
-                        }
+                if (collectionTypeKind is CollectionExpressionTypeKind.ArrayInterface ||
+                    hasSpreadElements)
+                {
+                    // Verify the existence of the List<T> members that may be used in lowering, even
+                    // though not all will be used for any particular collection expression. Checking all
+                    // gives a consistent behavior, regardless of collection expression elements.
+                    _ = _binder.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_List_T__ctor, _diagnostics, syntax: syntax);
+                    _ = _binder.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_List_T__ctorInt32, _diagnostics, syntax: syntax);
+                    _ = _binder.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_List_T__Add, _diagnostics, syntax: syntax);
+                    _ = _binder.GetWellKnownTypeMember(WellKnownMember.System_Collections_Generic_List_T__ToArray, _diagnostics, syntax: syntax);
+                }
 
-                        (collectionCreation, collectionBuilderMethod, collectionBuilderElementsPlaceholder) = bindCollectionBuilderInfo(
-                            this, node, targetType, collectionBuilderMethods, diagnostics);
-                        if (collectionCreation is null || collectionBuilderMethod is null || collectionBuilderElementsPlaceholder is null)
-                        {
-                            return BindCollectionExpressionForErrorRecovery(node, targetType, inConversion: true, diagnostics);
-                        }
+                // From this point out, all the remaining collection types end up converting all their elements
+                // to their actual element type and passing those along.  
```

Comment:
micro-nit: trim spaces.
```suggestion
                // to their actual element type and passing those along.
```